File: Semester 2/Assignment01/Linklist.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None
        self.last = None

    def add(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            self.last = new_node
        else:
            self.last.next = new_node
            self.last = new_node

    def addFirst(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
        if self.last is None:
            self.last = new_node

    def addLast(self, data):
        self.add(data)

    def clear(self):
        self.head = None
        self.last = None

    def getFirst(self):
        if self.head:
            return self.head.data
        else:
            raise IndexError("getFirst from empty list")

    def getLast(self):
        if not self.last:
            raise IndexError("getLast from empty list")
        return self.last.data

    def peekFirst(self):
        return self.getFirst()

    def peekLast(self):
        return self.getLast()

    def __str__(self):
        nodes = []
        current = self.head
        while current:
            nodes.append(str(current.data))
            current = current.next
        return " -> ".join(nodes)


# Example usage:
ll = LinkedList()
ll.add(1)
ll.addFirst(0)
ll.addLast(2)
print("Linked List:", ll)
print("First Element:", ll.getFirst())
print("Last Element:", ll.getLast())
ll.clear()
print("Linked List after clear:", ll)

File: Semester 2/Assignment01/Linklist.cpp
#include <iostream>

class Node
{
public:
    int data;
    Node *next;
};

// Global pointers for the head and last nodes
Node *head = nullptr;
Node *last = nullptr;

// Function to add a node to the end of the list
void add(int data)
{
    Node *new_node = new Node;
    new_node->data = data;
    new_node->next = nullptr;
    if (!head)
    {
        head = new_node;
        last = new_node;
    }
    else
    {
        last->next = new_node;
        last = new_node;
    }
}

// Function to add a node to the beginning of the list
void addFirst(int data)
{
    Node *new_node = new Node;
    new_node->data = data;
    new_node->next = head;
    head = new_node;
    if (!last)
    {
        last = new_node;
    }
}

// Function to add a node to the end of the list (alias for add)
void addLast(int data)
{
    add(data);
}

// Function to clear the list
void clear()
{
    Node *current = head;
    while (current)
    {
        Node *next = current->next;
        delete current;
        current = next;
    }
    head = nullptr;
    last = nullptr;
}

// Function to get the first element in the list
int getFirst()
{
    if (!head)
    {
        std::out_of_range("getFirst from empty list");
    }
    return head->data;
}

// Function to get the last element in the list
int getLast()
{
    if (!last)
    {
        std::out_of_range("getLast from empty list");
    }
    return last->data;
}

// Function to peek the first element in the list
int peekFirst()
{
    return getFirst();
}

// Function to peek the last element in the list
int peekLast()
{
    return getLast();
}

// Function to print the list
void display()
{
    Node *current = head;
    while (current)
    {
        std::cout << current->data;
        current = current->next;
        if (current)
        {
            std::cout << " -> ";
        }
    }
    std::cout << std::endl;
}

// Main function to demonstrate the usage of the linked list
int main()
{
    add(1);
    addFirst(0);
    addLast(2);
    add(3);
    std::cout << "Linked List: ";
    display();
    std::cout << "First Element: " << getFirst() << std::endl;
    std::cout << "Last Element: " << getLast() << std::endl;
    clear();
    std::cout << "Linked List after clear: ";
    display();

    return 0;
}
